dask_expr/categorical.py

```diff
@@ -0,0 +1,129 @@
+import functools
+
+from dask.dataframe.utils import (
+    AttributeNotImplementedError,
+    clear_known_categories,
+    has_known_categories,
+)
+from dask.utils import M
+
+from dask_expr.acessor import Accessor, PropertyMap
+from dask_expr.expr import Blockwise
+
+
+class CategoricalAccessor(Accessor):
+    """
+    Accessor object for categorical properties of the Series values.
+
+    Examples
+    --------
+    >>> s.cat.categories  # doctest: +SKIP
+
+    Notes
+    -----
+    Attributes that depend only on metadata are eager
+
+    * categories
+    * ordered
+
+    Attributes depending on the entire dataset are lazy
+
+    * codes
+    * ...
+
+    So `df.a.cat.categories` <=> `df.a._meta.cat.categories`
+    So `df.a.cat.codes` <=> `df.a.map_partitions(lambda x: x.cat.codes)`
+    """
+
+    _accessor_name = "cat"
+    _accessor_methods = (
+        "add_categories",
+        "as_ordered",
+        "as_unordered",
+        "remove_categories",
+        "rename_categories",
+        "reorder_categories",
+        "set_categories",
+    )
+    _accessor_properties = ()
+
+    @property
+    def known(self):
+        """Whether the categories are fully known"""
+        return has_known_categories(self._series)
+
+    def as_known(self, **kwargs):
+        """Ensure the categories in this series are known.
+
+        If the categories are known, this is a no-op. If unknown, the
+        categories are computed, and a new series with known categories is
+        returned.
+
+        Parameters
+        ----------
+        kwargs
+            Keywords to pass on to the call to `compute`.
+        """
+        if self.known:
+            return self._series
+        from dask_expr.collection import new_collection
+
+        categories = (
+            new_collection(PropertyMap(self._series.expr, "cat", "categories"))
+            .unique()
+            .compute()
+        )
+        return self.set_categories(categories.values)
+
+    def as_unknown(self):
+        """Ensure the categories in this series are unknown"""
+        if not self.known:
+            return self._series
+
+        from dask_expr import new_collection
+
+        return new_collection(AsUnknown(self._series.expr))
```

Comment:
Hrm, it's odd I think for Expr objects to return collections.  I think of Expr operations as being closed around Exprs.  Maybe some of this API belongs at the collection level and some of it belongs at the Expr level.  Maybe that's awkward though.  Thoughts?

Reply:
I am viewing the Accessors more as collections than as expressions. The accessor itself is registered on the Series collection